### DISCUSSION TITLE
Questions about some Linux handlers

### DISCUSSION THREAD

**Discussion by tm6907** [cmu-sei/GHOSTS#386] [Q&A] [answered]
Hi, I just have a few questions about some of json examples for different handlers that I'm trying to implement for a linux client. Apologies, I'm still learning JSON.

1. Here on the example for an SSH timeline,  what is the meaning of this section "<an IP>|<unique_key_from_credentials>|" Am I to fill the blanks with the values for a box to perform these SSH tasks on?
<img width="1077" alt="image" src="https://github.com/user-attachments/assets/624d1e28-2d82-43cb-954e-dbb5ba79fc70">

2. For the example handler for cron, does the "ScheduleType" of Cron affect the handler in anyway or is just running the command described in "TimelineEvents" ? Because it seems the "Schedule" doesn't affect when the Cron job is run, instead its determined by the "DelayAfter" value.
<img width="1077" alt="image" src="https://github.com/user-attachments/assets/d1b6330a-ab24-4447-8d89-3d2056a6225b">

3. Is there an example timeline block for the Curl handler on linux? 

4. Is the Bash handler for Linux clients the equivalent to the Command handler? If so should I use Bash over Command and is there an example timeline block? 

Thanks :)

**Comment by vramULB** [comment]
I'm not working directly on this project, but I hope this can help you with the first question:

You need to create a JSON file to add the credentials for the SSH connection, following the example below:

```json

{
  "Version": "1.0",
  "Data": {
    "credkey1": {"username": "user1", "password": "pw1base64"}, #Credentials for ip 192.168.20.1
    "credkey2": {"username": "user2", "password": "pw2base64"},  #Credentials for ip 192.168.20.2
  }
}
```
Then, add the path of the file to the CredentialsFile variable to be loaded. The JSON file will be deserialized into a C# object called Credentials. See more details here: [Credentials.cs.
](https://github.com/cmu-sei/GHOSTS/blob/master/src/Ghosts.Client/Infrastructure/Credentials.cs)
Next, the CommandArgs represents the commands for SSH. For an SSH connection, you need to retrieve the IP, username, and password of the target machine. Using the previous configuration, replace _<unique_key_from_credentials>_ with _credkey1_ as in the example above and add the commands you would like to be executed in the target machine:

```json

"CommandArgs": [
  "192.168.20.1 | credkey1 | ls -lah; ls -ltrh; help; pwd; date; time; uptime; uname -a; df -h; cd ~; cd /home/kali; touch cve1234.rb"
]
```

You can learn more about the running process here: [Ssh.cs.](https://github.com/cmu-sei/GHOSTS/blob/master/src/ghosts.client.linux/Handlers/Ssh.cs)

**Reply by tm6907** [reply]
Brilliant, thank you :)

**Reply by tm6907** [reply]
On this SSH handler, how would I check it has worked? The logs says its began and ssh command but not much else before it starts another handler. Nothing to see on the other VM I had it ssh to as well

**Comment by sei-dupdyke** [comment, marked as answer]
> 2... does the "ScheduleType" of Cron affect the handler in anyway or is just running the command described in "TimelineEvents" ? Because it seems the "Schedule" doesn't affect when the Cron job is run, instead its determined by the "DelayAfter" value.

Some wanted to be able to schedule specific activities at very specific times, so we implemented Quartz - a library that enabled cron style task scheduling (Quartz has others as well, but we felt most people would know cron).

I'd have to look into the specific issue you're seeing. TBH, we mostly use ghosts for randomized activity, and so very rarely if ever use cron (I'd actually argue to remove it, unless folks have compelling arguments to keep it).

> 3... Is there an example timeline block for the Curl handler on linux?

No, sorry, but a curl should run like any other command under bash.

> 4... Is the Bash handler for Linux clients the equivalent to the Command handler? If so should I use Bash over Command and is there an example timeline block?

Yes, bash for linux, Cmd for windows. There should be [samples here](https://github.com/cmu-sei/GHOSTS/tree/a4f30766092051cd48226c3f0cef533d8bbc6f11/src/Ghosts.Client/Sample%20Timelines).

**Reply by tm6907** [reply]
It seems for the Command handler on Linux it works anyway
